### PR TITLE
net: lib: azure_fota: Add job ID handling to avoid repeated update

### PR DESCRIPTION
**Pull request by jtguggedal**
Add job handling of job ID to avoid repeatedly downloading the same
firmware image.

In certain situations, the device can get into a loop
where it concinues to download the same firmware image over and
over again without being able to apply it, spending large amounts of
power and data traffic.
This patch prevents that from happening by reporting the job ID that
was last processed and comparing it to the ID of any incoming firmware
update.

Fixes CIA-222

### CODE REVIEW

File: subsys/net/lib/azure_fota/azure_fota.c [hunk 1]
```diff
@@ -176,20 +199,24 @@ static void fota_dl_handler(const struct fota_download_evt *fota_evt)
...
  * the device twin. If no action is needed, false is returned.
  */
 static bool parse_reported_status(const char *msg)
 {
 	struct cJSON *root_obj, *reported_obj, *fw_obj, *fw_status_obj,
-		*fw_current_obj, *fw_pending_obj;
+		*fw_current_obj, *fw_pending_obj, *job_id_obj;
 	bool report_needed = false;
 
 	/* The 'reported' object has the excpected structure if it exists:
```
Comment (coderbyheart) [inline subsys/net/lib/azure_fota/azure_fota.c:211]: Add a comment here, that jobs are considered new if no jobId is present, and that this is so we can support older shadows without breaking.

File: subsys/net/lib/azure_fota/azure_fota.c [hunk 2]
```diff
@@ -385,6 +429,23 @@ static int extract_fw_details(const char *msg)
 		goto clean_exit;
 	}
 
+	/* Firmware update job ID */
+	job_id_obj = cJSON_GetObjectItemCaseSensitive(fw_obj, "jobId");
+	if (job_id_obj == NULL) {
+		LOG_WRN("No jobId object found");
```
Comment (coderbyheart) [inline subsys/net/lib/azure_fota/azure_fota.c:435]:
```suggestion
		LOG_WRN("No jobId found");
```

File: subsys/net/lib/azure_fota/azure_fota.c [hunk 3]
```diff
@@ -216,6 +243,23 @@ static bool parse_reported_status(const char *msg)
...
```
Comment (coderbyheart) [inline subsys/net/lib/azure_fota/azure_fota.c:249]:
```suggestion
		LOG_WRN("No jobId found");
```

**Review by coderbyheart — CHANGES_REQUESTED** [review]
Also change occurences of `excpected` to `expected`.

File: subsys/net/lib/azure_fota/azure_fota.c (hunk 1 above)
Comment (jtguggedal) [reply subsys/net/lib/azure_fota/azure_fota.c:352]: Claryfying (hopefully) comments added where the job ID object is used

File: subsys/net/lib/azure_fota/azure_fota.c (hunk 3 above)
Comment (jtguggedal) [reply subsys/net/lib/azure_fota/azure_fota.c]: Fixed for both

**Review by umapraseeda — APPROVED** [review]

File: subsys/net/lib/azure_fota/azure_fota.c [hunk 4]
```diff
@@ -23,52 +23,78 @@
...
+static enum internal_state {
+	/* Library is uninitialized, no messages can be processed. */
+	STATE_UNINIT,
+	/* Initialized and waiting for incoming messages to process. */
+	STATE_INIT,
+	/* Downloading FOTA image. */
+	STATE_DOWNLOADING,
+} internal_state = STATE_UNINIT;
+
+/* Enum to keep the reported fota status */
```
Comment (simensrostad) [inline subsys/net/lib/azure_fota/azure_fota.c:60]: nit: keep track of?

File: subsys/net/lib/azure_fota/azure_fota.c [hunk 5]
```diff
@@ -23,52 +23,78 @@
...
+static struct fota_object {
 	char host[CONFIG_AZURE_FOTA_HOSTNAME_MAX_LEN];
 	char path[CONFIG_AZURE_FOTA_FILE_PATH_MAX_LEN];
 	char version[CONFIG_AZURE_FOTA_VERSION_MAX_LEN];
+	char job_id[CONFIG_AZURE_FOTA_JOB_ID_MAX_LEN];
 	size_t fragment_size;
-};
+} fota_object;
 
 /* Static variables */
```
Comment (simensrostad) [inline subsys/net/lib/azure_fota/azure_fota.c:80]: This line can be moved to include ```internal_state```, ```fota_status```, and ```fota_object```. I guess they are also static variables since they are declared in this case.

File: subsys/net/lib/azure_fota/azure_fota.c [hunk 6]
```diff
@@ -82,10 +108,51 @@ static char current_version[CONFIG_AZURE_FOTA_VERSION_MAX_LEN] =
 BUILD_ASSERT(0, "CONFIG_AZURE_FOTA_SEC_TAG must be defined");
 #endif
```
Comment (simensrostad) [inline subsys/net/lib/azure_fota/azure_fota.c:110]: Maybe add comments to signify where the static and internal interface begins? In the spirit of the comment present above that introduces the static variable section. This is out of scope I guess.

File: subsys/net/lib/azure_fota/azure_fota.c [hunk 7]
```diff
@@ -82,10 +108,51 @@ static char current_version[CONFIG_AZURE_FOTA_VERSION_MAX_LEN] =
...
+
+static enum fota_status reported_status_get(const char *status)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(rep_status_str); i++) {
+		if (strcmp(rep_status_str[i], status) == 0) {
+			return i;
+		}
+	}
+
+	return REP_STATUS_COUNT;
```
Comment (simensrostad) [inline subsys/net/lib/azure_fota/azure_fota.c:147]: Could we introduce an ```_INVALID``` status that is returned if the reported status is not present in the ```rep_status_str``` list? I guess the effect is the same, so no biggie.

File: subsys/net/lib/azure_fota/azure_fota.c [hunk 8]
```diff
@@ -175,21 +266,98 @@ static void fota_dl_handler(const struct fota_download_evt *fota_evt)
...
-	/* The 'reported' object has the excpected structure if it exists:
+	/* The 'reported' object has the expected structure if it exists:
 	 *	properties.reported.firmware : {
-			"fwUpdateStatus": <status>,
-			"currentFwVersion": <current firmware version>,
-			"pendingFwVersion": <FOTA version in progress (if any)>
+	 *		"fwUpdateStatus": <status>,
+	 *		"currentFwVersion": <current firmware version>,
+	 *		"pendingFwVersion": <FOTA version in progress (if any)>,
+	 *		"jobId": <FOTA update ID, formatted according to
```
Comment (simensrostad) [inline subsys/net/lib/azure_fota/azure_fota.c:357]: Does the ```AZURE_FOTA``` actually check if the jobID is in a format according to RFC4122? If not, it may be a little misleading to write that it is expected.

**Review by simensrostad — COMMENTED** [review]
Looks good. A few comments here and there.

File: subsys/net/lib/azure_fota/azure_fota.c (hunk 7 above)
Comment (jtguggedal) [reply subsys/net/lib/azure_fota/azure_fota.c]: Yeah, initially had that, but removed as it's not strictly needed as you mention. But will re-introduce it for clarity

File: subsys/net/lib/azure_fota/azure_fota.c (hunk 8 above)
Comment (jtguggedal) [reply subsys/net/lib/azure_fota/azure_fota.c:360]: It doesn't, so that reference is mainly to make sure users don't have larger strings than 32 hex chars + 4x `-`. Can be changed if desired

**Comment by jtguggedal** [comment]
> Looks good. A few comments here and there.

@simensrostad Comments have been addressed

**Review by simensrostad — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Comment by jtguggedal** [comment]
@coderbyheart Are you okay with the changes so that we can merge?

**Review by coderbyheart — APPROVED** [review]
Cursory review looks good.
Not tested manually.
